### PR TITLE
England Loneliness 2020/21 from the Community Life Survey

### PR DESCRIPTION
**Pull request by jennajt**
- Added notebook and csv using the Community Life Survey
- Mapped from Output Area Classification to LSOA

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -47,3 +47,15 @@ po/*~
 
 # RStudio Connect folder
 rsconnect/
+
+# lookup files
+*.feather
+
+# temporary working files
+cls.ods
+oac_lookup.csv
```
Comment (MikeJohnPage) [inline .gitignore]: Without access to these temporary files, I am unable to replicate your analysis. Can we not add these into `inst/` somewhere?

**Comment by MikeJohnPage** [comment]
In the notebook, you replace missing values with zero like so:

Screenshot 2023-07-28 at 15 20 23

This creates a cluster of points around zero, which weren't previously present in the data.

Screenshot 2023-07-28 at 15 26 07

If we look at the notes in the raw data for why there are `NA` values present for this data it says:

> Where the total number of responses is less than 30, this data has been suppressed.

This means we do not know the true values of the missing data. I would argue that replacing them with zero distorts the data in a negative way. Better, to leave them as missing, or impute a value.

**Comment by MikeJohnPage** [comment]
In this step when you select the cols `lsoa11_code` and `oa11_code`, you need to take the distinct values as they are repeated many times due to the different postcodes:

Screenshot 2023-07-28 at 15 51 47

The averaging performed in a later step should cancel out these repeat values, but it makes it harder to follow as-is.

**Comment by MikeJohnPage** [comment]
I think something has gone wrong here. The number of 2011 LSOA's in England should be 32,844, not 41,729. It looks like you have retained data for the devolved nations:

Screenshot 2023-07-28 at 16 28 53

**Comment by MikeJohnPage** [comment]
The file `../../extdata/lookup_21.feather` doesn't exist in the repo and should be shared:

Screenshot 2023-07-28 at 16 32 17

**Comment by MikeJohnPage** [comment]
Unfortunately, this approach isn't valid:

Screenshot 2023-07-28 at 16 36 43

This is because it assumes LSOA's are in a 1:1 relationship from 2011 to 2021, which is not true. See [this](https://github.com/britishredcrosssociety/health-inequalities-explorer/pull/50#issuecomment-1653959140) comment in a HIE PR which explains this problem in more detail. Unfortunately, the solution I have written to manage this, only works when aggregating from 2021 to 2011 LSOAs. To get the solution to work in this context (going from 2011 to 2021), I think the strategy needs inverting.

**Review by MikeJohnPage — CHANGES_REQUESTED** [review]

**Comment by MikeJohnPage** [comment]
Thanks for all your hard work on this, it is appreciated. I've left you comments / things to change above. I had bit of a hard time reasoning about some of the Python code, so I replicated the analysis in R to follow along as I was doing the review, fixing a few of the issues I have described above (e.g., not replacing the missing values with zero, keeping England only LSOA's). I added a file with this R code to this branch [here](https://github.com/humaniverse/loneliness/blob/england_cls/inst/python/notebooks/england_2021.R). Feel free to build off this for the LSOA aggregations if helpful or continue making the changes in the notebook.

Thanks 😃

**Comment by jennajt** [comment]
Your comments above have been addressed. In addition I restructured the repo to make the code reproducible:

-  The main script to review is `cls_england_2020.py `. It contains everything needed to download all the files and process the script to produce a csv. Note this script calls an R script. I haven't added an R environment for it, let me know if you think it is required.
- The notebook is not for reviewing - it is where the experimentation is done.
- I've created a `requirements.txt` document to manage python dependencies. You will need to create your own virtual env and install the libraries and versions, instructions in `README.md`
- I've created an RDA file and some basic documentation for the dataset.

**Comment by MikeJohnPage** [comment]
Amazing work. Merging now. 🚀

**Pull request merged**